src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,289 @@
+      local ihl = ip4_header:ihl() * 4
+      assert(ihl == get_ihl(pkt))
+      assert(ihl == ip4_header:sizeof())
+      assert(total_length - ihl >= 0)
+      assert(total_length == pkt.length - eth_proto:sizeof())
+      return total_length - ihl
+   end
+
+   local function pkt_frag_offset(pkt)
+      assert(pkt.length >= (eth_proto:sizeof() + ip4_proto:sizeof()))
+      local ip4_header = ip4_proto:new_from_mem(pkt.data + eth_proto:sizeof(),
+                                                pkt.length - eth_proto:sizeof())
+      return ip4_header:frag_off() * 8
+   end
+
+   local function pkt_total_length(pkt)
+      assert(pkt.length >= (eth_proto:sizeof() + ip4_proto:sizeof()))
+      local ip4_header = ip4_proto:new_from_mem(pkt.data + eth_proto:sizeof(),
+                                                pkt.length - eth_proto:sizeof())
+      return ip4_header:total_length()
+   end
+
+   local function check_packet_fragment(orig_pkt, frag_pkt, is_last_fragment)
+      -- Ethernet fields
+      local orig_hdr = eth_proto:new_from_mem(orig_pkt.data, orig_pkt.length)
+      local frag_hdr = eth_proto:new_from_mem(frag_pkt.data, frag_pkt.length)
+      assert(orig_hdr:src_eq(frag_hdr:src()))
+      assert(orig_hdr:dst_eq(frag_hdr:dst()))
+      assert(orig_hdr:type() == frag_hdr:type())
+
+      -- IPv4 fields
+      orig_hdr = ip4_proto:new_from_mem(orig_pkt.data + eth_proto:sizeof(),
+                                        orig_pkt.length - eth_proto:sizeof())
+      frag_hdr = ip4_proto:new_from_mem(frag_pkt.data + eth_proto:sizeof(),
+                                        frag_pkt.length - eth_proto:sizeof())
+      assert(orig_hdr:ihl() == frag_hdr:ihl())
+      assert(orig_hdr:dscp() == frag_hdr:dscp())
+      assert(orig_hdr:ecn() == frag_hdr:ecn())
+      assert(orig_hdr:ttl() == frag_hdr:ttl())
+      assert(orig_hdr:protocol() == frag_hdr:protocol())
+      assert(orig_hdr:src_eq(frag_hdr:src()))
+      assert(orig_hdr:dst_eq(frag_hdr:dst()))
+
+      assert(pkt_payload_size(frag_pkt) == frag_pkt.length - eth_proto:sizeof() - ip4_proto:sizeof())
+
+      if is_last_fragment then
+         assert(band(frag_hdr:flags(), 0x1) == 0x0)
+      else
+         assert(band(frag_hdr:flags(), 0x1) == 0x1)
+      end
+   end
+
+   -- Packet with 1200 bytes of payload
+   local pkt = assert(make_ipv4_packet(1200))
+
+   -- MTU bigger than the packet size
+   local code, result = assert(fragment_ipv4(pkt, 1500))
+   assert(code == FRAGMENT_UNNEEDED)
+   assert(pkt == result)
+
+   -- Keep a copy of the packet, for comparisons
+   local orig_pkt = packet.allocate()
+   orig_pkt.length = pkt.length
+   ffi.copy(orig_pkt.data, pkt.data, pkt.length)
+
+   assert(pkt.length > 1200, "packet short than payload size")
+
+   code, result = assert(fragment_ipv4(pkt, 1000))
+   assert(code == FRAGMENT_OK)
+   assert(#result == 2,
+          "fragmentation returned " .. tostring(#result) .. " packets (2 expected)")
```

Comment:
Concatenating a number to a string automatically converts the number to a string, no need to explicitly coerce it.